### PR TITLE
Kernel: Donate time slices for socket interactions

### PR DESCRIPTION
**Pull request by gunnarbeutner**
This PR implements support for intelligently scheduling threads for IPC interactions:

1. It adds support to the kernel that lets it remember which user threads access a file via read()/write()/select()/poll().
2. Fixes a scheduler hang with non-runnable threads:

When a thread that is non-runnable tried to donate their remaining time slice to another thread that was also non-runnable we'd remove the thread from the ready queue but didn't add it back.

This would cause the thread to never get scheduled again. Previously this could not happen because non-runnable threads had no way of donating their time slice to anyone else.

3. Donates time slices to peer when waiting on a socket:

This causes the user-mode thread to donate their remaining time slice to the peer of a socket when it does a blocking `read()`/`write()` or waits for data with `select()`/`poll()`.

4. Adds the `donate_peer` system call:

This system call donates the calling thread's remaining time slice to the peer of the specified file descriptor (only sockets are
supported for now).

`LibIPC` has been updated to use this after writing a complete IPC message to a socket.

5. Adds statistics for time slices donated/received. These show up in `SystemMonitor`. Note, that `SystemMonitor` does not currently support threads so you might have to look at `/proc/all` instead for some processes.

This might be biased and quite subjective but Super Mario's keyboard input seems significantly more response than it was before. I'd suggest trying this in comparison with the `master` branch.

Screenshot from 2021-05-05 05-10-50

### CODE REVIEW

File: Kernel/ThreadBlockers.cpp [hunk 1]
```diff
@@ -436,6 +441,13 @@ void Thread::SelectBlocker::was_unblocked(bool did_timeout)
     }
 }
 
+RefPtr<Thread> Thread::SelectBlocker::pick_donate_target() const
+{
+    if (m_fds.is_empty())
+        return nullptr;
+    return m_fds[0].description->file().likely_peer_thread();
```
Comment (alimpfard) [inline Kernel/ThreadBlockers.cpp:448]: Why the first one?
Why not the second one, or the most starved one, or the least scheduled one, or why not just distribute equally?

Comment (alimpfard) [reply Kernel/ThreadBlockers.cpp:448]: (that is to say, please at least describe your reasoning for picking that specific fd as a comment above that, so people won't be stuck thinking "is there a specific reason" and will just see the specific reason)

File: Userland/Libraries/LibIPC/Connection.h [hunk 2]
```diff
@@ -101,6 +101,10 @@ class Connection : public Core::Object {
             total_nwritten += nwritten;
         }
 
+#ifdef __serenity__
+        (void)donate_peer(m_socket->fd());
```
Comment (bgianfo) [inline Userland/Libraries/LibIPC/Connection.h:105]: nitpick: should this be a separate commit, it's not mentioned in the title / description.

File: Userland/Libraries/LibC/unistd.cpp [hunk 3]
```diff
@@ -681,6 +681,12 @@ int donate(int tid)
     __RETURN_WITH_ERRNO(rc, rc, -1);
 }
 
+int donate_peer(int fd)
+{
+    int rc = syscall(SC_donate_peer, fd);
```
Comment (bgianfo) [inline Userland/Libraries/LibC/unistd.cpp:686]: nitpick: Add `+ LibC` to the commit title?

File: Kernel/Syscalls/sched.cpp [hunk 4]
```diff
@@ -29,6 +30,21 @@ KResultOr<int> Process::sys$donate(pid_t tid)
     return 0;
 }
 
+KResultOr<int> Process::sys$donate_peer(int fd)
+{
+    REQUIRE_PROMISE(stdio);
```
Comment (bgianfo) [inline Kernel/Syscalls/sched.cpp:35]: The description claims only sockets are supported, should this be `REQUIRE_PROMISE_FOR_SOCKET_DOMAIN(...)` or even directly `REQUIRE_PROMISE(unix)`?

Also we don't seem to validate that the description is a socket anywhere?

**Comment by awesomekling** [comment]
>LibIPC has been updated to use this after writing a complete IPC message to a socket.

This feels like it should only be done for synchronous messages, no?

**Comment by gunnarbeutner** [comment]
> > LibIPC has been updated to use this after writing a complete IPC message to a socket.
> 
> This feels like it should only be done for synchronous messages, no?

That makes a lot of sense. In which case the `donate_peer` system call kind of loses its purpose because the client would be blocking on `select()` again. I think I might just remove it.
